README.md
# Phantom Desktop Wallet

![Phantom Desktop Wallet](./banner.png)

[![Build Status](https://badgen.now.sh/github/status/XPHChain/desktop-wallet)](https://github.com/XPHChain/desktop-wallet/actions?query=branch%3Adevelop)
[![Latest Version](https://badgen.now.sh/github/release/XPHChain/desktop-wallet)](https://github.com/XPHChain/desktop-wallet/releases)
[![License: MIT](https://badgen.now.sh/badge/license/MIT/green)](https://opensource.org/licenses/MIT)

> Lead Maintainer: [Lúcio Rubens](https://github.com/luciorubeens)

## Download

[Latest Release](https://github.com/XPHChain/ark-desktop/releases/latest)

## Installing via Package Managers

### Arch Linux

Install via [AUR](https://aur.archlinux.org/packages/ark-desktop):

> Manjaro

```shell
pamac build ark-desktop
```

### Mac OS X

Install via [Homebrew](https://brew.sh/):

```shell
brew cask install ark-desktop-wallet
```

## Translations

Translations are no longer accepted for the Desktop Wallet. We are using plugins to provide languages.

## Development

### Requirements

#### Ubuntu
In Ubuntu the development files of `libudev` are necessary:
```
sudo apt-get install libudev-dev libusb-1.0-0-dev
```

#### Windows
- Python 2.7
- Visual Studio 2017

#### Node 12
To download, head over to [here](https://nodejs.org/en/) and download Node 12.

If you already have npm installed, you can run
```
npm install -g n
sudo n 12
```

#### Yarn
Install the Yarn dependency manager
```
npm install -g yarn
```

### Commands

<details><summary>List of commands</summary>

``` bash
# Install dependencies
yarn install

# Execute the application. Making changes in the code, updates the application (hot reloading).
yarn dev

# Lint all JS/Vue files in the `src` and `__tests__`
yarn lint

# Lint, and fix, all JS/Vue files in `src` and `__tests__`
yarn lint:fix

# Collect the code and produce a compressed file
yarn pack

# Build electron application for production (Current OS)
yarn build

# Build electron application for production (Windows)
yarn build:win

# Build electron application for production (Mac)
yarn build:mac

# Build electron application for production (Linux)
yarn build:linux

# Run unit and end-to-end tests
yarn test

# Run unit tests
yarn test:unit

# Run unit tests and generate and display the coverage report
yarn test:unit:coverage

# Run unit tests and watch for changes to re-run the tests
yarn test:unit:watch

# Run end-to-end tests, without building the application
yarn test:e2e

# Build the application and run end-to-end tests
yarn test:e2e:full
```

</details>

## Security

If you discover a security vulnerability within this package, please send an e-mail to security@ark.io. All security vulnerabilities will be promptly addressed.

## Credits

This project exists thanks to all the people who [contribute](../../contributors).

## License

[MIT](LICENSE) © [Phantom Ecosystem](https://ark.io)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## Unreleased

## Version 2.2.1 (2019-01-30)

Broadcasting to multiple peers fixed and improved usability for custom networks 🐛

### Added

- Different network name style & add ledger flags (#991)
- Allow expanding and collapsing the wallet/contact sidebar (#1015)
- Persistent sorting in wallet/contact table (#1002)
- Use network static fees as max (#1022)

### Changed

- Make wallet heading name more visible in dark mode (#996)
- Adjust announcement section to fit new design (#982)
- New instructions section for wallet import/new and network overview page (#1003)
- New wallet selection styles and ui (#1006)
- Success/error colors and delay on clipboard button tooltip (#1010)
- Adapt the contact creation page to new style (#1012)
- Remove unnecessary delegate api calls (#995)
- Differentiate between wallet and contact rename (#988)
- Inconsistencies between network versions (#1016)
- Add sidebar animation (#1021)
- Loading screen when broadcasting to many peers (#1023)
- New selection design for the network overview (#1037)

### Fixed

- Correctly handle zero balance on wallet heading info (#1000)
- Conditional alignment of items (#1005)
- Do not shake when generating new wallets (#1008)
- Validation on sign/verify modals & signed messages alignment (#1014)
- Remove the beta warning and collpase list of commands in README
- Add forged to delegate model (#1017)
- Replace the profile avatars with a new component `ProfileAvatar` (#1019)
- Importing of profile avatar selection modal component (#1031)
- Invert identicon hover behaviour (#1025)
- Errors during and after removing profiles (#1033)
- Show "Unvote" action on delegate tab when using Ledger (#1032)
- Allow selecting the balance on the wallet page heading (#994)
- Broadcasting modal not closing on unvote (#1038)
- Update @arkecosystem/client (#1036)
- Pasting or writing numbers with colon (",") on the send transfer inputs (#1009)

## Previous versions

See [GitHub releases](https://github.com/XPHChain/desktop-wallet/releases).


src/renderer/services/synchronizer.js
import { pullAll } from 'lodash'
import { flatten } from '@/utils'
import { announcements, fees, ledger, market, peer, wallets } from './synchronizer/'
/**
 * This class adds the possibility to define actions (not to confuse with Vuex actions)
 * that could be dispatched using 2 modes: `default` and `focus`.
 *
 * Each mode has an interval configuration that establishes the delay until the
 * next update:
 *  - `focus` should be used on sections that display that kind of data to users
 *  - `default` should be used to check if new data is available
 *
 * There is also a way to pause and unpause the synchronization that is useful
 * for not executing actions when the users do not need fresh data.
 */
export default class Synchronizer {
  get intervals () {
    // Phantom block production time
    const block = 8000

    const intervals = {
      longest: block * 300,
      longer: block * 100,
      medium: block * 25,
      shorter: block * 10,
      shortest: block * 3,
      block,
      // Number of milliseconds to wait to evaluate which actions should be run
      loop: 2000
    }

    return intervals
  }

  get config () {
    const { loop, shortest, shorter, medium, longer, longest } = this.intervals

    const config = {
      announcements: {
        default: { interval: longest, delay: loop * 6 },
        focus: { interval: medium }
      },
      delegates: {
        default: { interval: longer, delay: loop * 3 },
        focus: { interval: longer }
      },
      fees: {
        default: { interval: null },
        focus: { interval: shorter }
      },
      ledgerWallets: {
        default: { interval: shorter },
        focus: { interval: shortest }
      },
      market: {
        default: { interval: medium },
        focus: { interval: shorter }
      },
      peer: {
        default: { interval: longer },
        focus: { interval: shorter }
      },
      wallets: {
        default: { interval: shorter },
        focus: { interval: shortest }
      }
    }
    config.contacts = config.wallets

    return config
  }

  get $client () {
    return this.scope.$client
  }

  get $store () {
    return this.scope.$store
  }

  /**
   * @param {Object} config
   * @param {Vue} config.scope - Vue instance that would be synchronized
   */
  constructor ({ scope }) {
    this.scope = scope
    this.actions = {}
    this.focused = []
    this.paused = []
  }

  /**
   * Define an action that would be called later periodically
   * @param {String} actionId
   * @param {Object} config
   * @param {Function} actionFn
   */
  define (actionId, config, actionFn) {
    if (typeof actionFn !== 'function') {
      throw new Error('[$synchronizer] action is not a function')
    }
    ;['default', 'focus'].forEach(mode => {
      const { interval } = config[mode]
      if (!interval && interval !== null) {
        throw new Error(`[$synchronizer] \`interval\` for \`${mode}\` mode should be a Number bigger than 0 (or \`null\` to ignore it)`)
      }
    })

    this.actions[actionId] = {
      calledAt: 0,
      isCalling: false,
      fn: actionFn,
      ...config
    }
  }

  /**
   * Focus on these actions: instead of refreshing their data on the normal pace,
   * change to the `focus` frequency.
   *
   * Focusing on 1 or several actions, unfocused the rest
   * @params {(...String|Array)} actions - ID of the actions to focus on
   */
  focus (...actions) {
    this.focused = flatten(actions)
    this.unpause(this.focused)
  }

  /**
   * Add 1 or more actions to additionally focus
   * @params {(...String|Array)} actions - ID of the actions to focus on
   */
  appendFocus (...actions) {
    this.focused = flatten([actions, this.focused])
    this.unpause(actions)
  }

  /**
   * Remove action from focus
   * @params {(...String|Array)} actions - ID of the actions to focus on
   */
  removeFocus (...actions) {
    pullAll(this.focused, flatten(actions))
  }

  /**
   * Pause these actions. They would not be dispatched until they are unpaused
   * or focused
   * @params {(...String|Array)} actions - ID of the actions to pause
   */
  pause (...actions) {
    this.paused = flatten(actions)
  }

  /**
   * Enable these paused actions again
   * @params {(...String|Array)} actions - ID of the actions to unpause
   */
  unpause (...actions) {
    pullAll(this.paused, flatten(actions))
  }

  /**
   * Trigger these actions 1 time.
   * As a consequence the interval of those actions is updated.
   * @params {(...String|Array)} actions - ID of the actions to unpause
   */
  trigger (...actions) {
    flatten(actions).forEach(actionId => this.call(actionId))
  }

  /**
   * Invoke the action and update the last time it has been called, when
   * it has finished its sync or async execution
   * @param {String} actionId
   */
  async call (actionId) {
    const action = this.actions[actionId]
    if (!action) {
      return
    }

    action.isCalling = true
    await action.fn()
    action.calledAt = (new Date()).getTime()
    action.isCalling = false
  }

  /*
   * Starts to dispatch the actions periodically
   */
  ready () {
    /**
     * Run all the actions
     */
    const run = (options = {}) => {
      Object.keys(this.actions).forEach(actionId => {
        if (!this.paused.includes(actionId)) {
          const action = this.actions[actionId]

          if (!action.isCalling) {
            if (options.immediate) {
              this.call(actionId)
            } else {
              const mode = this.focused.includes(actionId) ? 'focus' : 'default'
              const { interval } = action[mode]

              // A `null` interval means no interval, so the action does not run
              if (interval !== null) {
                // Delay the beginning of the periodic action run
                if (!action.calledAt && action[mode].delay) {
                  action.calledAt += action.delay
                }

                const nextCallAt = action.calledAt + interval
                const now = (new Date()).getTime()

                if (nextCallAt <= now) {
                  this.call(actionId)
                }
              }
            }
          }
        }
      })
    }

    const runLoop = () => {
      // Using `setTimeout` instead of `setInterval` allows waiting to the
      // completion of async functions
      setTimeout(() => {
        run()
        runLoop()
      }, this.intervals.loop)
    }

    // Run the first time
    run({ immediate: true })
    runLoop()
  }

  defineAll () {
    this.define('announcements', this.config.announcements, async () => {
      await announcements(this)
    })

    // TODO focus on contacts only (currently wallets and contacts are the same)
    // this.define('contacts', this.config.contacts, async () => {
    //   console.log('defined CONTACTS')
    // })

    // NOTE: not used currently
    // this.define('delegates', this.config.delegates, async () => {
    //   await delegates(this)
    // })

    this.define('fees', this.config.fees, async () => {
      await fees(this)
    })

    this.define('market', this.config.market, async () => {
      await market(this)
    })

    this.define('peer', this.config.peer, async () => {
      await peer(this)
    })

    // TODO allow focusing on 1 wallet alone, while using the normal mode for the rest
    this.define('wallets', this.config.wallets, async () => {
      await wallets(this)
    })

    this.define('wallets:ledger', this.config.ledgerWallets, async () => {
      await ledger(this)
    })
  }
}


config/index.js
/* eslint-disable @typescript-eslint/no-var-requires */
const os = require('os')
const path = require('path')

exports.APP = {
  website: 'https://phantomchain.org',
  transactionExpiryMinutes: 45
}

exports.NETWORKS = [
  require('./networks/phantom.json'),
  require('./networks/mainnet.json'),
  require('./networks/devnet.json')
]

exports.PEERS = {
  'phantom.mainnet': require('./peers/phantom.json'),
  'ark.mainnet': require('./peers/mainnet.json'),
  'ark.devnet': require('./peers/devnet.json')
}

exports.ANNOUNCEMENTS = {
  rssUrl: 'https://blog.ark.io/feed'
}

exports.VENDOR_FIELD = {
  defaultMaxLength: 64
}

exports.I18N = {
  defaultLocale: 'en-US',
  enabledLocales: [
    'en-US'
  ]
}

exports.BIP39 = {
  defaultLanguage: 'english',
  languages: [
    'chinese_simplified',
    'chinese_traditional',
    'english',
    'french',
    'italian',
    'japanese',
    'korean',
    'spanish'
  ]
}

exports.TRANSACTION_GROUPS = {
  STANDARD: 1,
  MAGISTRATE: 2
}

exports.TRANSACTION_TYPES = {
  MULTI_SIGN: -1,

  GROUP_1: {
    TRANSFER: 0,
    SECOND_SIGNATURE: 1,
    DELEGATE_REGISTRATION: 2,
    VOTE: 3,
    MULTI_SIGNATURE: 4,
    IPFS: 5,
    MULTI_PAYMENT: 6,
    DELEGATE_RESIGNATION: 7,
    HTLC_LOCK: 8,
    HTLC_CLAIM: 9,
    HTLC_REFUND: 10
  },

  GROUP_2: {
    BUSINESS_REGISTRATION: 0,
    BUSINESS_RESIGNATION: 1,
    BUSINESS_UPDATE: 2,
    BRIDGECHAIN_REGISTRATION: 3,
    BRIDGECHAIN_RESIGNATION: 4,
    BRIDGECHAIN_UPDATE: 5,
    ENTITY: 6
  }
}

exports.TRANSACTION_TYPES_ENTITY = {
  TYPE: {
    BUSINESS: 0,
    PRODUCT: 1,
    PLUGIN: 2,
    MODULE: 3,
    DELEGATE: 4
  },

  SUBTYPE: {
    NONE: 0
  },

  ACTION: {
    REGISTER: 0,
    UPDATE: 1,
    RESIGN: 2
  }
}

exports.INTERVALS = {
  short: 30000, // 30 seconds
  medium: 60000, // 1 minute
  long: 180000 // 3 minute
}

exports.MARKET = {
  defaultCurrency: 'BTC',
  crypto: [
    'BTC',
    'ETH',
    'LTC'
  ],
  currencies: {
    BTC: { symbol: 'Ƀ', fractionDigits: 8 },
    ETH: { symbol: 'Ξ', fractionDigits: 8 },
    LTC: { symbol: 'Ł', fractionDigits: 8 },
    AUD: { symbol: 'A$', fractionDigits: 2 },
    BRL: { symbol: 'R$', fractionDigits: 2 },
    CAD: { symbol: 'C$', fractionDigits: 2 },
    CHF: { symbol: 'CHF', fractionDigits: 2 },
    CNY: { symbol: '¥', fractionDigits: 2 },
    EUR: { symbol: '€', fractionDigits: 2 },
    GBP: { symbol: '£', fractionDigits: 2 },
    HKD: { symbol: 'HK$', fractionDigits: 2 },
    IDR: { symbol: 'IDR', fractionDigits: 2 },
    INR: { symbol: '₹', fractionDigits: 2 },
    JPY: { symbol: '¥', fractionDigits: 0 },
    KRW: { symbol: '₩', fractionDigits: 0 },
    MXN: { symbol: 'MX$', fractionDigits: 2 },
    RUB: { symbol: '₽', fractionDigits: 2 },
    USD: { symbol: '$', fractionDigits: 2 }
  }
}

// This section handles fetching Desktop-Wallet Plugins from the NPM registry.
// It should remain "ARK" unless intentionally implementing a custom package.
exports.PLUGINS = {
  adapters: ['npm'],
  pluginsUrl: 'https://raw.githubusercontent.com/ark-ecosystem-desktop-plugins/config/master/plugins.json',
  categories: [
    'gaming',
    'theme',
    'language',
    'utility',
    'other'
  ],
  devPath: path.resolve(os.homedir(), '.ark-desktop/plugins-dev'),
  maxKeywords: 5,
  keywords: [
    '@arkecosystem',
    'desktop-wallet',
    'plugin'
  ],
  officialScope: 'arkecosystem',
  officialAuthor: 'ARK Ecosystem',
  path: path.resolve(os.homedir(), '.ark-desktop/plugins'),
  reportUrl: 'https://ark.io/contact',
  sharePath: path.resolve(os.homedir(), '.ark-desktop/share'),
  updateInterval: {
    value: 1,
    unit: 'day'
  },
  validation: require('./plugin-validation.json')
}

exports.THEMES = [
  {
    id: 'light',
    title: 'Light theme' // TODO translate
  },
  {
    id: 'dark',
    title: 'Dark theme' // TODO translate
  }
]

exports.V1 = {
  fees: {
    GROUP_1: [
      0.1 * 1e8, // Transfer
      5 * 1e8, // Second signautre
      25 * 1e8, // Delegate registration
      1 * 1e8, // Vote
      5 * 1e8, // Multisignature
      5 * 1e8, // IPFS
      1 * 1e8, // Multi-payment
      25 * 1e8, // Delegate resignation
      1 * 1e8, // HTLC Lock
      0 * 1e8, // HTLC Claim
      0 * 1e8 // HTLC Refund
    ],

    GROUP_2: [
      50 * 1e8, // Business Registration
      50 * 1e8, // Business Resignation
      50 * 1e8, // Business Update
      50 * 1e8, // Bridgechain Registration
      50 * 1e8, // Bridgechain Resignation
      50 * 1e8 // Bridgechain Update
    ]
  }
}
